### ISSUE TITLE
TreeViewer automatically selects first row if first click is child expansion elsewhere in the list

### ISSUE THREAD

**Issue by rkostes** [eclipse-platform/eclipse.platform.swt#314]
**Describe the bug**
A TreeView is opened.  There is no selection.  All of the rows have children so they all have an expansion arrow on the left.  
Clicking any expansion arrow expands the arrow to show the children, BUT, it also automatically selects the first row in the entire list.
Since a selection in this list causes other windows to update, it is confusing for the user because the updates in the other window
are related to the first item in the list and not what they clicked on.

**To Reproduce**
Create a TreeViewer with 3 rows and each row has a child.  Open the view and select the 2d or 3rd arrow.  Make sure you 
select the arrow and not the row itself.  The selected arrow will expand but the first row in the list will get selected.

**Expected behavior**
The expected behavior is that the selected arrow expands but nothing should be selected.

**Screenshots**
![Capture](https://user-images.githubusercontent.com/3661171/184454783-4ee4886e-42bb-4b95-a35c-d14538bf5fb4.JPG)

In the screenshot ConnectionDiscovery was selected and it expanded as expected.  But ADVA_LAG, the first item in the list, got selected.  Keep in mind that this only occurs when the widget is first opened.  Selecting subsequent arrows, the first item
remains selected.  Once you actually select something, then the widget behaves properly.

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [] All OS
- - [ x] Windows
- - [ ] Linux
- - [ ] macOS 

2. Additional OS info (e.g. OS version, Linux Desktop, etc)
Windows 10

3. JRE/JDK version
 12.0.1

**Version since**
 4.1 and 4.25

I notice that on contraction of an arrow, the widget deals with the case where an item disappearing is the one that was selected.  I am guessing that on expansion it wants to make sure the selection is correctly highlighted and since there isn't one, it just picks the first row.

**Comment by Phillipus** [comment]
I can't reproduce this behaviour in our RCP app, in Eclipse or in the TreeViewer snippet. Can you provide a test snippet that shows this issue?

**Comment by rkostes** [comment]
So I started removing all my code to get it to bare minimum.  This is what I found.  I do not understand it.

		me.setLayout(new GridLayout(1, false));
		constructWidgetsEntry();
		constructWidgetsTree();

me is a Composite.  constructWidgetsEntry creates a set of widgets in a Composite above the tree.  In the screenshot above you can see the bottom border just above the tree.

If I comment out constuctWidgetsEntry(), the tree behaves fine.  If I call constructWidgetsEntry(), I get the bad behavior in the tree.  Baffling to me so I stripped constructWidgetsEntry() down to one line.

	    Composite entry = new Composite(me, SWT.BORDER);

Somehow creating this blank composite, causes the bad behavior in the tree.  

Here is the stripped down tree code.

	private void constructWidgetsTree()
	{
		tree = new Tree(me, SWT.BORDER | SWT.H_SCROLL | SWT.V_SCROLL | SWT.FULL_SELECTION);
		
		tree.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
		tree.setHeaderVisible(true);
		tree.setLinesVisible(true);		
		//tree.addSelectionListener(this);
		//tree.addMouseListener(this);		
	
		this.treeStandardColumn(tree, TREE_NAME, "Folder/Dump Name", 380);
		//this.treeStandardColumn(tree, TREE_NOTES, "N", 20, "Notes attached?").setResizable(false);
		//this.treeStandardColumn(tree, TREE_NUM_FEATURES, "#F", 30, "Number of Features").setResizable(false);
		//lastColumn = this.treeStandardColumn(tree, TREE_VENDOR, "Vendor", 100);
				
	}
	
	public void reload()
	{
		this.fillTree();
		//mdm.containerSelected(null);
	}
	

	private void fillTree()
	{
		tree.removeAll();

		TreeItem folderItem = null;
		TreeItem leafItem = null;
		
		folderItem = new TreeItem(tree, 0);	// The roots of the tree
		folderItem.setFont(fontBold());
		folderItem.setText(TREE_NAME, "FOLDER 1");
		
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 1");
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 2");

		folderItem = new TreeItem(tree, 0);	// The roots of the tree
		folderItem.setFont(fontBold());
		folderItem.setText(TREE_NAME, "FOLDER 2");
		
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 1");
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 2");
		
		folderItem = new TreeItem(tree, 0);	// The roots of the tree
		folderItem.setFont(fontBold());
		folderItem.setText(TREE_NAME, "FOLDER 3");
		
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 1");
		leafItem = new TreeItem(folderItem, 0);
		leafItem.setText(TREE_NAME, "leaf 2");

	}

**Comment by Phillipus** [comment]
Perhaps another control has the focus, and then the tree gets the focus when expanding a node.

Here's a test snippet that shows the tree getting the focus when expanding a node:

<details>
<summary>Click to reveal code...</summary>

```java
import org.eclipse.jface.layout.GridDataFactory;
import org.eclipse.jface.layout.TreeColumnLayout;
import org.eclipse.jface.viewers.CellLabelProvider;
import org.eclipse.jface.viewers.ColumnWeightData;
import org.eclipse.jface.viewers.ITreeContentProvider;
import org.eclipse.jface.viewers.TreeViewer;
import org.eclipse.jface.viewers.TreeViewerColumn;
import org.eclipse.jface.viewers.ViewerCell;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class TestViewer {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("TreeViewer");
        shell.setSize(600, 400);
        shell.setLayout(new GridLayout(1, false));
        
        Composite entry = new Composite(shell, SWT.BORDER);
        GridDataFactory.create(GridData.FILL_HORIZONTAL).hint(SWT.DEFAULT, 20).applyTo(entry);
        entry.setFocus();
        
        Composite treeComposite = new Composite(shell, SWT.BORDER);
        treeComposite.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
        TreeColumnLayout layout = new TreeColumnLayout();
        treeComposite.setLayout(layout);
        
        final TreeViewer tv = new TreeViewer(treeComposite, SWT.H_SCROLL | SWT.V_SCROLL | SWT.FULL_SELECTION);

        TreeViewerColumn column1 = new TreeViewerColumn(tv, SWT.NONE, 0);
        column1.getColumn().setText("c 1");
        layout.setColumnData(column1.getColumn(), new ColumnWeightData(33, true));
        
        TreeViewerColumn column2 = new TreeViewerColumn(tv, SWT.NONE, 1);
        column2.getColumn().setText("c 2");
        layout.setColumnData(column2.getColumn(), new ColumnWeightData(33, true));

        TreeViewerColumn column3 = new TreeViewerColumn(tv, SWT.NONE, 2);
        column3.getColumn().setText("c 3");
        column3.getColumn().setMoveable(true);
        layout.setColumnData(column3.getColumn(), new ColumnWeightData(33, true));

        tv.setContentProvider(new MyContentProvider());
        tv.setLabelProvider(new MyLabelProvider());
        
        tv.getTree().setHeaderVisible(true);
        tv.getTree().setLinesVisible(true);
        
        tv.setInput("");

        shell.open();
        
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) display.sleep();
        }
        
        display.dispose();
    }
    
    static class MyContentProvider implements ITreeContentProvider {
        @Override
        public Object[] getElements(Object inputElement) {
            return new Object[] {"One", "Two", "Three"};
        }

        @Override
        public Object[] getChildren(Object parentElement) {
            return new Object[] {"Hello", "There"};
        }

        @Override
        public Object getParent(Object element) {
            return null;
        }

        @Override
        public boolean hasChildren(Object element) {
            return true;
        }
    }
    
    static class MyLabelProvider extends CellLabelProvider {
        @Override
        public void update(ViewerCell cell) {
            String text = (String)cell.getElement();
            cell.setText(text);
        }
    }
}
```
</details>

**Comment by Phillipus** [comment]
If you change the line in my snippet where the Tree is created to add `SWT.MULTI` then the Tree doesn't gain the focus when expanding:

```
final TreeViewer tv = new TreeViewer(treeComposite,
                                               SWT.H_SCROLL | SWT.V_SCROLL | SWT.FULL_SELECTION | SWT.MULTI);
```

**Comment by rkostes** [comment]
Yes, I added SWT.MULTI and the tree behaves fine.  However, I do not want a multi-select tree.

And yes, if I call tree.setFocus() immediately, the window opens with the first row selected.

I am no GUI expert, but isn't that a bug?  Why should the widget getting focus cause a selection event?  If a button gets focus, it does not consider itself pressed.  Is there anything I can do to prevent this?

**Comment by Phillipus** [comment]
> I am no GUI expert, but isn't that a bug? Why should the widget getting focus cause a selection event? If a button gets focus, it does not consider itself pressed. Is there anything I can do to prevent this?

I don't know if it's a bug or how to prevent this. Maybe someone else knows? (Just trying to help pin the issue down)

**Comment by rkostes** [comment]
Thanks for your help.  I think you have certainly narrowed it down to something very specific.  I never would have gotten here on my own.

**Comment by rkostes** [comment]
I created a workaround for this bug by adding a TreeListener to catch the expand event and various flags to know that the first selection event coming in is bogus and to set the doIt flag to false in the event.  The first row still gets highlighted but the selection is ignored.

What got me on to this solution was fixing another strange behavior.  I have a search widget with a entry field, Find and Find Next buttons.  I captured the ENTER in the entry field with a TraverseListener, I do the find and then setFocus to the Find Next button all in this method.  However, the Find Next button was also generating a Selection event.  I had to set the doIt flag to false in the TraverseListener handling method.  Seems like when I changed focus to a new widget, the new widget got the ENTER event.
